In [1]:
# imort libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import path of data
path = r'C:\Users\nikik\17-12-2025 Instacart Basket Analysis'

In [3]:
# Import data, Orders csv
df_ords = pd.read_csv(os.path.join(path, 'Data','Prepared Data', 'orders_wrangled.csv'))
# poroducts csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv' ), index_col = False)

# 1. Checking consistancy for df_prods

In [4]:
#check for consistency
df_prods.describe() 

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Check for mixed-type data

In [5]:
# Create a Dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_test['mix'].dtype

dtype('O')

# Missing values

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# There are 16 missing values in product name field. These are likely due to incomplete records in the Instacart system, for example when certain products were not fully labeled at the time of data entry.

# Address the missing values using an appropriate method

In [14]:
# creating a new dataframe dropping the missing values
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

# I created a new dataframe by dropping rows with missing values. This was chosen because the missing data was minimal and removing it ensures a complete dataset for analysis.

# Check for duplicate values.

In [18]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# A few duplicate rows were found, likely due to repeated product entries in the system. Removing them ensures each product is represented only once.

# Addressing Duplicates

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

# I removed duplicate rows using drop_duplicates() to ensure each product appears only once and avoid distortions in the analysis.

# 2. Run the df.describe() function on your df_ords dataframe. 

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [24]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# All min–max values look reasonable. order_hour_of_day correctly ranges 0–23 and days_since_prior_order ranges 0–30 as expected. The only issue is fewer records in days_since_prior_order, indicating missing values that should be investigated (likely first orders).

# Check for mixed-type data in your df_ords dataframe

In [25]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# I checked all columns in df_ords for mixed data types, and no columns returned any results. This means there are no mixed data types present, and the dataframe is consistent in terms of data types

# 5. Run a check for missing values in your df_ords dataframe

In [26]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# There are 206209 missing values in order_since_prior_order field. this is because of these correspond to customers’ first orders, which have no previous order to calculate a value from. All other columns have complete data.

# 6. Address the missing values using an appropriate method.

In [27]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [28]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# I filled the missing values in days_since_prior_order with 0 because these missing values correspond to customers’ first orders, which naturally have no prior order. This approach preserves all rows in the dataset while keeping the data logically consistent. Optionally, a flag column can be added to explicitly mark first orders for analysis.

# 7. Run a check for duplicate values in your df_ords data.

In [29]:
df_dups = df_ords[df_ords.duplicated()]

In [30]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [31]:
df_ords.shape

(3421083, 7)

# I checked df_ords for duplicate rows. No duplicates were found, which indicates that each order in the dataset is unique and the data is consistent.

In [32]:
# Exporting cleaned data
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# I exported the final cleaned dataframes to CSV files in the “Prepared Data” folder. 
- products_cleaned.csv  contains the cleaned product data. 
- orders_cleaned.csv contains the cleaned orders data, with missing values handled.  
These files are ready for analysis.